### PR TITLE
Handle broken contracts

### PR DESCRIPTION
**Pull request by turbolent**
Work towards #3480 

## Description

Implements the program recovery as described in FLIP https://github.com/onflow/flips/pull/283

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/ft_test.go [hunk 1]
```diff
@@ -766,3 +768,263 @@ func BenchmarkRuntimeFungibleTokenTransfer(b *testing.B) {
 
 	utils.RequireValuesEqual(b, nil, mintAmountValue, sum)
 }
+
+// TODO:
```
Comment (turbolent) [inline runtime/ft_test.go:772]: The goal is to uncomment and use this version of the ExampleToken and have the transaction still work

**Review by SupunS — APPROVED** [review]
Nice!

File: runtime/ft_test.go [hunk 2]
```diff
@@ -766,3 +770,408 @@ func BenchmarkRuntimeFungibleTokenTransfer(b *testing.B) {
...
+	vaultDeclaration := contractDeclaration.Members.CompositesByIdentifier()[fungibleTokenVaultTypeIdentifier]
+	if vaultDeclaration == nil {
+		return false
+	}
+
+	// Check if the Vault resource has a balance field
+	balanceFieldDeclaration := getField(vaultDeclaration, fungibleTokenVaultTypeBalanceFieldName)
+	if balanceFieldDeclaration == nil {
+		return false
+	}
```
Comment (SupunS) [inline runtime/ft_test.go:909]: Do we need to check these? Can we assume if a contract is deployed, then it must be a valid contract (and hence, given it conforms to the FungibleToken, it should have all these fields)?

File: runtime/ft_test.go [hunk 3]
```diff
@@ -766,3 +770,408 @@ func BenchmarkRuntimeFungibleTokenTransfer(b *testing.B) {
...
+	const transaction = `
+      import FungibleToken from 0x1
+      import ExampleToken from 0x1
+
+      transaction {
+
+          let vault: @ExampleToken.Vault
+
+          prepare(signer: auth(LoadValue) &Account) {
+              self.vault <- signer.storage.load<@ExampleToken.Vault>(from: /storage/exampleTokenVault)!
```
Comment (SupunS) [inline runtime/ft_test.go:1147]: Could we add some more operations to the test? e.g:
- getting the type (`self.vault.getType()`)
- casting to `FungibleToken.Vault`
- Storing the read value back to storage
- Calling a function on the `self.vault`
- etc.

File: runtime/ft_test.go (hunk 2 above)
Comment (turbolent) [reply runtime/ft_test.go:909]: Yes, I guess we could maybe make these checks more lenient. Not 100% sure if that could introduce loopholes

File: runtime/ft_test.go (hunk 3 above)
Comment (turbolent) [reply runtime/ft_test.go:1147]: Good idea! Will add these 👍

Comment (turbolent) [reply runtime/ft_test.go:1151]: Added in [7197c69](https://github.com/onflow/cadence/pull/3482/commits/7197c6908705c98e8cc791c342efb2da6cfe9c72) and [acd782a](https://github.com/onflow/cadence/pull/3482/commits/acd782a9a053181c46194063913a0bbb81f46abb)

**Comment by turbolent** [comment]
For anyone looking at this PR: This PR only adds support for the program recovery functionality as a runtime interface function. It does NOT provide an implementation. An implementation of the function is only provided for demonstratation / testing purposes in a test case. Merging the PR does not change any behaviour for Cadence programs.

**Pull request merged**